go/login/server/server.go

```diff
@@ -0,0 +1,178 @@
+// limitations under the License.
+
+// Package server implements GLOME-login server framework.
+package server
+
+import (
+	"encoding/hex"
+	"fmt"
+	"net/http"
+	"sync"
+
+	"../../login" // Modify when it is pushed
+)
+
+const (
+	// MaxResponseSize is the maximum size in characaters of the response token
+	MaxResponseSize = 44 // 32 bytes base64 encoded
+)
+
+var (
+	// ErrInvalidResponseLen denotes that response length provided is invalid
+	ErrInvalidResponseLen = fmt.Errorf("invalid response length provided")
+)
+
+// Authorizer responds to an authorization request. The method
+// GrantLogin returns whether an user is allowed to run an action in a host, and a error value.
+//
+// Some considerations need to be held while implementing this interface:
+// - Allow should accept an empty string as command, denoting that
+// an empty action is passed in the request.
+// - If no user can be obtained from request metadata, an empty string is to be
+// passed as default value.
+// - Both hostIDType and hostID can be empty. Whether this refer to a default value
+// or not is to be user configurable.
+type Authorizer interface {
+	GrantLogin(user string, hostID string, hostIDType string, action string) (bool, error)
+}
+
+// AuthorizerFunc type is an adapter to allow the use of ordinary functions as an Authorizer.
+type AuthorizerFunc func(user string, hostID string, hostIDType string, action string) (bool, error)
+
+// GrantLogin calls a(user, hostID, hostIDType, action)
+func (a AuthorizerFunc) GrantLogin(user string, hostID string, hostIDType string, action string) (bool, error) {
+	return a(user, hostID, hostIDType, action)
+}
+
+// LoginServer is a framework for glome-login servers, Creation of the object is managed via
+// the NewLoginServer method. To create one it is necessary to provide an authorizer. Optionally, a series
+// of options can be added at creation.
+type LoginServer struct {
+	// Keys manages the keys used by the server.
+	Keys *KeyManager
+
+	// Unexported fields.
+	auth        Authorizer
+	authLock    sync.RWMutex
+	loginParser *login.Server
+
+	responseLen uint8
+	userHeader  string
+}
+
+// Authorizer sets Authorizer function for LoginServer in a concurrentsafe way.
+func (s *LoginServer) Authorizer(a Authorizer) {
+	s.authLock.Lock()
+	s.auth = a
+	s.authLock.Unlock()
+}
+
+// NewLoginServer creates a new server with provided Authorizer and, optionally, selected options.
+func NewLoginServer(a Authorizer, options ...func(*LoginServer) error) (*LoginServer, error) {
+	srv := LoginServer{
+		auth:        a,
+		Keys:        NewKeyManager(),
+		responseLen: MaxResponseSize,
+	}
+	srv.loginParser = srv.newLoginParser()
+
+	for _, option := range options {
+		if err := option(&srv); err != nil {
+			return nil, err
+		}
+	}
+
+	return &srv, nil
+}
+
+// ResponseLen is an option to be provided to NewServer on creation. Its sets responseLen to
+// provided length. Raise error if provided length is not in 0-MaxResponseSize
```

Comment:
I belive the purpose of `responseLen` is not documented, so referring to it does not explain why I would want to use `ResponseLen`.

Reply:
Fixed, now refer to the size of the response token